mlir/lib/Dialect/Rock/Transforms/RemoveRedundantCasts.cpp

```diff
@@ -0,0 +1,510 @@
+    auto loc = input->getLoc();
+    auto genericOp = rewriter.create<linalg::GenericOp>(
+        loc,
+        /*resultTypes=*/TypeRange{outputType},
+        /*inputs=*/ValueRange{input->getResult(0)},
+        /*outputs=*/ValueRange{output->getResult(0)},
+        /*indexingMaps=*/indexingMaps,
+        /*iteratorTypes=*/iteratorTypes,
+        /*doc=*/"",
+        /*library_call=*/"",
+        /*bodyBuild=*/
+        [](OpBuilder &builder, Location loc, ValueRange args) {
+          Value blockArg = args[0];
+          Value outputArg = args[1];
+          Type oType = outputArg.getType();
+          Value truncResult =
+              builder.create<arith::TruncFOp>(loc, oType, blockArg);
+          builder.create<linalg::YieldOp>(loc, truncResult);
+        });
+
+    return genericOp;
+  }
+
+  bool handleRockGemmWrapper(Operation *rockOp, linalg::GenericOp extGen,
+                             PatternRewriter &rewriter) const {
+    LLVM_DEBUG(llvm::dbgs()
+               << "\tFound "
+               << "RockGemmWrapperInterface/RockGemmGemmWrapperInterface: "
+               << *rockOp << "\n");
+
+    Value extGenResult = extGen.getResult(0);
+    Value extGenInput = extGen.getInputs()[0];
+
+    auto extOutputType = dyn_cast<RankedTensorType>(extGenResult.getType());
+    auto extInputType = dyn_cast<RankedTensorType>(extGenInput.getType());
+
+    if (!extOutputType || !extInputType) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\tOne or more types are not RankedTensorType, skipping\n");
+      return false;
+    }
+
+    Type rockOutputElementType;
+    Value rockOutputOp;
+    if (auto rockI = dyn_cast<RockGemmWrapperInterface>(rockOp)) {
+      rockOutputElementType = rockI.getCType();
+      rockOutputOp = rockI.getOutArgument()->get();
+    } else if (auto rockGemmI =
+                   dyn_cast<RockGemmGemmWrapperInterface>(rockOp)) {
+      rockOutputElementType = rockGemmI.getCType();
+      rockOutputOp = rockGemmI.getOutArgument()->get();
+    } else {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\tNot a RockGemmWrapperInterface, skipping\n");
+      return false;
+    }
+
+    auto rockOutputType = dyn_cast<RankedTensorType>(rockOutputOp.getType());
+    if (!rockOutputType) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\tOutput type is not RankedTensorType, skipping\n");
+      return false;
+    }
+
+    if (rockOutputType.getElementType() != extInputType.getElementType()) {
+      LLVM_DEBUG(
+          llvm::dbgs()
+          << "\tTrunc input type doesn't match ext output type, or trunc "
+          << "output type doesn't match ext input type, skipping\n");
+      return false;
+    }
+
+    OpBuilder builder(rockOp->getContext());
+    builder.setInsertionPoint(rockOp);
+
+    // If this op uses mfma, it will accumulate in higher precision (F32 or I32)
+    auto features = rock::getFeatures(rockOp);
+    bool isMfma = bitEnumContainsAll(features, GemmFeatures::mfma);
```

Comment:
This should also work on wmma, it also accumulates in higher precision. 

Also check the non-accel path if it is doing accumulation in higher precision or not.